### DISCUSSION TITLE
Creating a new SQLAlchemy dialect

### DISCUSSION THREAD

**Discussion by am-kinetica** [sqlalchemy/sqlalchemy#11462] [Usage Questions] [answered]
We at Kinetica [Kinetica](https://www.kinetica.com) would like to create a new dialect for our database. We have a functional DBAPI v2 implementation that works in the sync mode for now, the async implementation is in works to be released soon. 
The questions that we have are:
1. What is exactly covered in the compliance suite of the dialects ? Is there a document I we could go through ?
2. Is it possible that a partial implementation of the dialect could be compliant, e.g., if we have a pass-through mode to the DBAPI enabled for now without compilations of SQL statements, would it be considered as compliant ?
3. Is it mandatory to have the dialect in a state so that all ORM constructs are working or is it possible to leave them out in the first cut and still have a compliant dialect ?

It would be a great help if I can get answers to these and have some documents to read through.

**Comment by zzzeek** [comment, marked as answer]
the document for creating new dialects is at https://github.com/sqlalchemy/sqlalchemy/blob/main/README.dialects.rst .    The range of behaviors and features that can be disabled for testing within the compliance suite is specific to the test cases you'd see in https://github.com/sqlalchemy/sqlalchemy/tree/main/lib/sqlalchemy/testing/suite and you can see what they are in https://github.com/sqlalchemy/sqlalchemy/blob/main/lib/sqlalchemy/testing/requirements.py .

it's not really possible to have a dialect that is "without compilations of SQL statements" since you're subclassing DefaultDialect which itself refers to a series of default SQL compiler classes.   if your database's interpretation of SQL is so different than the standard that even simple statements don't work, then that wouldn't be able to get through the compliance suite at all.

for ORM constructs, none of these link to a dialect directly, they all build upon Core, so the compliance suite should have no ORM features at all tested within it (I haven't checked if this is totally true but it should be near zero), so you wouldn't need to concern yourself with ORM behaviors at the dialect level, the Core SQL / execution / etc. tests should cover what's needed.

**Reply by CaselIT** [reply]
It may also be useful to take a look at other dialect to see how a sqlalchemy dialect is organized. You find a list here https://docs.sqlalchemy.org/en/20/dialects/index.html#external-dialects

Also feel free to send a PR to add yours to the table once you are done!

**Reply by am-kinetica** [reply]
Thanks for your replies, I will get back here once I am ready with the dialect.

**Reply by am-kinetica** [reply]
I have another question; is support for `setuptools` necessary or can we just use `poetry` ?

**Reply by CaselIT** [reply]
I think as long as the dialect installs correctly and the metadata is properly generated for sqlalchemy to find it, what you use to install it doesn't matter

**Comment by am-kinetica** [comment]
I am currently working on developing the Kinetica dialect and started testing it. The issue is that I am getting statements like:
`INFO:sqlalchemy.engine.Engine:ROLLBACK` in my test output. How do I stop emitting these implicit transaction outputs or how to get rid of implicit transactions completely ?

**Reply by CaselIT** [reply]
That's sqlalchemy calling rollback in the dbapi driver. Usually that happens when returning the connection to the pool. It's usually possible to call that method at any time.

Sqlalchemy does not use implicit transactions as far as I'm aware.

**Reply by am-kinetica** [reply]
Alright, so, if I want to control that, I need to check the implementation of the DBAPI that we have, is that it ?

**Reply by CaselIT** [reply]
I assumed you had looked at https://docs.sqlalchemy.org/en/20/core/internals.html#sqlalchemy.engine.Dialect.do_rollback

**Reply by am-kinetica** [reply]
The implementations that I have in our dialect are these:
`    def do_begin(self, dbapi_connection):
        # Disable automatic transactions by overriding the `do_begin` method to do nothing
        pass

    def do_commit(self, dbapi_connection):
        # Disable commits
        pass

    def do_rollback(self, dbapi_connection):
        # Disable rollback
        pass


`

**Reply by am-kinetica** [reply]
But these seem to have no effect.

**Reply by CaselIT** [reply]
I'm not sure what the issue is, sorry. Can you explain what you want to do and what does not work?

**Reply by am-kinetica** [reply]
I am testing a simple code with the dialect that I have developed:

```py
def test_create_table():
    from sqlalchemy import MetaData, Table, Column, Integer, String

    metadata_obj = MetaData()

    engine = create_engine(
        "kinetica://",
        connect_args={
            "url": URL,
            "username": USER,
            "password": PASS,
            "bypass_ssl_cert_check": BYPASS_SSL_CERT_CHECK,
        },
    )

    test_table1 = Table(
        "test_table_1",
        metadata_obj,
        Column("id", Integer, primary_key=True),
        Column("str_1", String(30)),
        Column("str_full", String),
        schema="sqlalchemy",
        quote=True,
        quote_schema=True,
    )
    metadata_obj.create_all(engine)
```

The output that I am getting is as follows:
```
INFO:sqlalchemy.engine.Engine:BEGIN (implicit)
INFO:sqlalchemy.engine.Engine:DROP SCHEMA IF EXISTS sqlalchemy CASCADE
INFO:sqlalchemy.engine.Engine:[generated in 0.00018s] ()
INFO:sqlalchemy.engine.Engine:CREATE SCHEMA sqlalchemy
INFO:sqlalchemy.engine.Engine:[generated in 0.00012s] ()
INFO:sqlalchemy.engine.Engine:BEGIN (implicit)
INFO:sqlalchemy.engine.Engine:SELECT 1 FROM KI_CATALOG.KI_OBJECTS WHERE UPPER(SCHEMA_NAME)=? AND UPPER(OBJECT_NAME)=? AND (OBJ_KIND = 'V' or OBJ_KIND = 'R') 
INFO:sqlalchemy.engine.Engine:[generated in 0.00013s] ('sqlalchemy', 'test_table_1')
INFO:sqlalchemy.engine.Engine:
CREATE  TABLE "sqlalchemy"."test_table_1" (
        id INTEGER NOT NULL, 
        str_1 VARCHAR(30), 
        str_full VARCHAR
) 

INFO:sqlalchemy.engine.Engine:[no key 0.00010s] ()
INFO:sqlalchemy.engine.Engine:COMMIT
INFO:sqlalchemy.engine.Engine:ROLLBACK
```
Now, my question is:

1. Why do I see things like **INFO:sqlalchemy.engine.Engine:BEGIN (implicit) Engine:COMMIT
 and Engine:ROLLBACK** when in my dialect I have turned them off.

**Reply by CaselIT** [reply]
That's the log of what sqlalchemy is doing. It's calling commit/rollback on the sqlalchemy connection. It's not sending the string commit/rollback to the database, since usually those operations are achieved by using the commit()/rollback() methods.

Looking at your logs this seems strange though 
```
INFO:sqlalchemy.engine.Engine:BEGIN (implicit)
INFO:sqlalchemy.engine.Engine:DROP SCHEMA IF EXISTS sqlalchemy CASCADE
INFO:sqlalchemy.engine.Engine:[generated in 0.00018s] ()
INFO:sqlalchemy.engine.Engine:CREATE SCHEMA sqlalchemy

**Reply by am-kinetica** [reply]
Could you please explain why it looks strange ? Maybe I am missing something.

**Reply by CaselIT** [reply]
There is no drop/create schema issued in your example, so it's strange that such statements are executed

**Reply by am-kinetica** [reply]
I see, that is being called from a method which is called prior to the method that I have described:

`def drop_and_recreate_schema(conn):
    conn.execute(text("DROP SCHEMA IF EXISTS sqlalchemy CASCADE"))
    conn.execute(text("CREATE SCHEMA sqlalchemy"))
`

**Reply by CaselIT** [reply]
Understood.

Since your db seems to not support transactions, it may make sense to tell sqlalchemy that autocommit is active and that's the only isolation level available. 

https://docs.sqlalchemy.org/en/20/core/internals.html#sqlalchemy.engine.default.DefaultDialect.get_default_isolation_level
https://docs.sqlalchemy.org/en/20/core/internals.html#sqlalchemy.engine.default.DefaultDialect.default_isolation_level
https://docs.sqlalchemy.org/en/20/core/internals.html#sqlalchemy.engine.default.DefaultDialect.get_isolation_level_values

Etc to indicate that autocommit is the only available one

**Reply by am-kinetica** [reply]
Right, our DB is non-transactional. Thanks for your pointers around this, I will update the dialect accordingly.

**Reply by am-kinetica** [reply]
I am trying to customize table creation in our dialect using this class:

```py
class CreateExternalTable(CreateTable):
    """_summary_

    Args:
        CreateTable (_type_): _description_
    """
    __visit_name__ = "create_table"

    def __init__(self, table, remote_query, file_paths=None, options=None, **kw):
        """_summary_

        Args:
            table (_type_): _description_
            remote_query (_type_): _description_
            file_paths (_type_, optional): _description_. Defaults to None.
            options (_type_, optional): _description_. Defaults to None.
        """
        super().__init__(table, **kw)
        self.remote_query = remote_query
        self.file_paths = file_paths
        self.options = options or {}
```

And I also have the the method `visit_create_table` overridden in my own `DDLCompiler` derived class. However, it seems the method from the SQLAlchemy class is only being invoked and the visit method that I have implemented is not having any effect. Any suggestions ?

**Reply by CaselIT** [reply]
How are you testing what method is getting invoked? Doing `str(statement)` will use the default compiler.
If you have overridden the `ddl_compiler` attribute of the class with your subclass, then you can use it by doing `statement.compile(dialet=your_dialect_instance)`

PS: when posting multi-line code snipped use 3 back-ticks: https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#quoting-code

**Reply by am-kinetica** [reply]
I have been going through the tests for the different dialects implemented by HANA, Snowflake and Rockset. What is curious is that the Rockset dialect has no tests associated. Is it not a mandate to have the tests with the dialect ? I can just see an `example.py`.

**Reply by CaselIT** [reply]
External dialect are external, there is no verification done by the sqlalchemy team to check compliance or anything like that.

**Reply by am-kinetica** [reply]
Ok, so, if we want to list our dialect on the page containing the links to the external dialects, what exactly do we need to do ? We do have enough test coverage as of now for our dialect to be usable.

**Reply by CaselIT** [reply]
Just send a pr adding the link to that page

**Reply by am-kinetica** [reply]
Alright, thanks.

**Reply by am-kinetica** [reply]
What is the best way to support a statement like:
```UPDATE  eb
SET     sal = e.sal, manager_id = e.manager_id
FROM    example.employee_backup eb, example.employee e
WHERE   eb.id = e.id
```

**Reply by am-kinetica** [reply]
Another form of the same statement that we support is:
```UPDATE  eb
SET     sal = e.sal, manager_id = e.manager_id
FROM    example.employee_backup eb
JOIN    example.employee e ON eb.id = e.id
```

**Reply by CaselIT** [reply]
Are those the update syntax? If so make the update render that kind of statement

**Reply by am-kinetica** [reply]
Thanks, we could handle these forms using custom generation of the statements.

Now, we are facing another issue;  we are getting warnings like :
```
sqlalchemy_api_examples.py:276: SAWarning: TypeDecorator UnsignedBigInteger() will not produce a cache key because the ``cache_ok`` attribute is not set to True.  This can have significant performance implications including some performance degradations in comparison to prior SQLAlchemy versions.  Set this attribute to True if this type object's state is safe to use in a cache key, or False to disable this warning. (Background on this warning at: https://sqlalche.me/e/20/cprf)
  result = conn.execute(stmt)
sqlalchemy_api_examples.py:276: SAWarning: TypeDecorator IPV4() will not produce a cache key because the ``cache_ok`` attribute is not set to True.  This can have significant performance implications including some performance degradations in comparison to prior SQLAlchemy versions.  Set this attribute to True if this type object's state is safe to use in a cache key, or False to disable this warning. (Background on this warning at: https://sqlalche.me/e/20/cprf)
  result = conn.execute(stmt)
```

**Reply by am-kinetica** [reply]
What I can't find from the docs is:
1. When and how to decide whether to set `cache_ok` to `True` or `False` ?

**Reply by CaselIT** [reply]
have you taken a look at the docs for cache_ok? https://docs.sqlalchemy.org/en/20/core/type_api.html#sqlalchemy.types.ExternalType.cache_ok

**Reply by am-kinetica** [reply]
Created the PR - https://github.com/sqlalchemy/sqlalchemy/pull/11895
